lib/SamplerUtils.cpp

```diff
@@ -13,19 +13,53 @@
 // limitations under the License.
 
 #include "SamplerUtils.h"
+#include "Types.h"
+#include "Constants.h"
 
 using namespace llvm;
 
 namespace clspv {
 
-Value *NormalizedCoordinate(Module &M, IRBuilder<> &B, Value *Coord, Value *Img,
-                            Type *ImgTy) {
+Value *GetImageDimFP(Module &M, IRBuilder<> &B, Value *Img, Type *ImgTy) {
+  auto float4Ty = FixedVectorType::get(B.getFloatTy(), 4);
   auto getImageSizesFct = M.getOrInsertFunction(
-      "clspv.get_image_sizes",
-      FunctionType::get(Coord->getType(), {ImgTy}, false));
+      "_Z13get_image_dim11ocl_image3d",
+      FunctionType::get(FixedVectorType::get(B.getInt32Ty(), 4), {ImgTy},
+                        false));
   Value *ImgSizes = B.CreateCall(getImageSizesFct, {Img});
+  return B.CreateSIToFP(ImgSizes, float4Ty);
+}
+
+Value *NormalizedCoordinate(Module &M, IRBuilder<> &B, Value *Coord,
+                            Value *ImgDimFP, bool FilteringNearest) {
+  if (FilteringNearest) {
+    auto float4Ty = FixedVectorType::get(B.getFloatTy(), 4);
+    auto getFloorFct = M.getOrInsertFunction(
+        "floor", FunctionType::get(float4Ty, {float4Ty}, false));
+    Coord = B.CreateCall(getFloorFct, {Coord});
+    Coord = B.CreateFAdd(Coord, ConstantFP::get(float4Ty, 0.5));
+  }
+
+  return B.CreateFDiv(Coord, ImgDimFP);
+}
 
-  return B.CreateFDiv(Coord, ImgSizes);
+bool isReadImage3DWithNonLiteralSampler(CallInst *call) {
+  DenseMap<Value *, Type *> cache;
+  auto Name = call->getCalledFunction()->getName();
+  if (Name.contains("read_image") && Name.contains("ocl_sampler")) {
```

Comment:
Prefer using the builtins lookup here. Also, I think infer type for the image is unnecessary since it is never a pointer anymore.

Reply:
Done